File: typescript/src/mcpClient.ts
import path from 'node:path';
import process from 'node:process';
import { fileURLToPath, pathToFileURL } from 'node:url';
import { Client } from '@modelcontextprotocol/sdk/client/index.js';
// import { StdioClientTransport } from "@modelcontextprotocol/sdk/client/stdio.js";
// --- STREAMABLE HTTP transport ---
// Uncomment to switch (also comment out the StdioClientTransport import above).
import { StreamableHTTPClientTransport } from '@modelcontextprotocol/sdk/client/streamableHttp.js';
import type { ProgressCallback } from '@modelcontextprotocol/sdk/shared/protocol.js';
import {
  CallToolResultSchema,
  CreateMessageRequestSchema,
  ListRootsRequestSchema,
  LoggingMessageNotificationSchema,
  type ClientCapabilities,
  type Tool,
  type Prompt,
  type PromptMessage,
  type CallToolResult,
  type CreateMessageRequest,
  type CreateMessageResult,
  type LoggingMessageNotification,
  type Root,
} from '@modelcontextprotocol/sdk/types.js';

export type SamplingHandler = (
  request: CreateMessageRequest,
) => Promise<CreateMessageResult>;

export type LoggingHandler = (notification: LoggingMessageNotification) => void;

// Converts path strings to Root objects. Equivalent to `_create_roots` in
// `mcp_client.py`.
function createRoots(rootPaths: string[]): Root[] {
  return rootPaths.map((rootPath) => {
    const resolved = path.resolve(rootPath);
    return {
      uri: pathToFileURL(resolved).href,
      name: path.basename(resolved) || 'Root',
    };
  });
}

// The MCP client. This is the TypeScript equivalent of `mcp_client.py`.
export class MCPClient {
  private command: string;
  private args: string[];
  private env?: Record<string, string>;
  private samplingHandler?: SamplingHandler;
  private loggingHandler?: LoggingHandler;
  private roots: Root[];
  private session: Client | null = null;
  // --- STDIO transport (default) ---
  // private transport: StdioClientTransport | null = null;
  // --- STREAMABLE HTTP transport ---
  // Comment out the line above and uncomment the line below to switch.
  private transport: StreamableHTTPClientTransport | null = null;

  constructor(
    command: string,
    args: string[],
    env?: Record<string, string>,
    samplingHandler?: SamplingHandler,
    loggingHandler?: LoggingHandler,
    roots?: string[],
  ) {
    this.command = command;
    this.args = args;
    this.env = env;
    this.samplingHandler = samplingHandler;
    this.loggingHandler = loggingHandler;
    this.roots = roots ? createRoots(roots) : [];
  }

  async connect(): Promise<void> {
    // --- STDIO transport (default) ---
    // this.transport = new StdioClientTransport({
    //   command: this.command,
    //   args: this.args,
    //   env: this.env,
    // });

    // --- STREAMABLE HTTP transport ---
    // Comment out the block above and uncomment the line below to switch.
    // Requires the server to be running with the streamable HTTP block
    // (see mcpServer.ts) and reachable at this URL.
    this.transport = new StreamableHTTPClientTransport(
      new URL('http://127.0.0.1:3000/mcp'),
    );

    const capabilities: ClientCapabilities = {};
    if (this.samplingHandler) capabilities.sampling = {};
    if (this.roots.length > 0) capabilities.roots = {};

    this.session = new Client(
      { name: 'mcp-chat-ts', version: '1.0.0' },
      Object.keys(capabilities).length > 0 ? { capabilities } : undefined,
    );

    if (this.samplingHandler) {
      this.session.setRequestHandler(
        CreateMessageRequestSchema,
        this.samplingHandler,
      );
    }

    if (this.loggingHandler) {
      this.session.setNotificationHandler(
        LoggingMessageNotificationSchema,
        (notification) => {
          this.loggingHandler?.(notification);
        },
      );
    }

    if (this.roots.length > 0) {
      const roots = this.roots;
      this.session.setRequestHandler(ListRootsRequestSchema, async () => ({
        roots,
      }));
    }

    await this.session.connect(this.transport);
  }

  getSession(): Client {
    if (this.session === null) {
      throw new Error('Client session not initialized. Call connect() first.');
    }
    return this.session;
  }

  async listTools(): Promise<Tool[]> {
    const result = await this.getSession().listTools();
    return result.tools;
  }

  async callTool(
    toolName: string,
    toolInput: Record<string, unknown>,
    onProgress?: ProgressCallback,
  ): Promise<CallToolResult | null> {
    const result = await this.getSession().callTool(
      { name: toolName, arguments: toolInput },
      CallToolResultSchema,
      onProgress ? { onprogress: onProgress } : undefined,
    );
    return result as CallToolResult;
  }

  async listPrompts(): Promise<Prompt[]> {
    const result = await this.getSession().listPrompts();
    return result.prompts;
  }

  async getPrompt(
    promptName: string,
    args: Record<string, string>,
  ): Promise<PromptMessage[]> {
    const result = await this.getSession().getPrompt({
      name: promptName,
      arguments: args,
    });
    return result.messages;
  }

  async readResource(uri: string): Promise<unknown> {
    const result = await this.getSession().readResource({ uri });
    const resource = result.contents[0];

    if (resource && 'text' in resource) {
      if (resource.mimeType === 'application/json') {
        return JSON.parse(resource.text);
      }
      return resource.text;
    }

    return null;
  }

  async cleanup(): Promise<void> {
    await this.session?.close();
    this.session = null;
    this.transport = null;
  }
}

// For testing
async function main() {
  // Spawns mcpServer.ts through the same Node binary using the tsx loader
  // (equivalent to Python's `command="uv", args=["run", "mcp_server.py"]`).
  const serverPath = fileURLToPath(new URL('./mcpServer.ts', import.meta.url));
  const client = new MCPClient(process.execPath, [
    '--import',
    'tsx',
    serverPath,
  ]);

  await client.connect();
  try {
    const result = await client.listTools();
    console.log(result);
  } finally {
    await client.cleanup();
  }
}

const isMainModule =
  process.argv[1] !== undefined &&
  fileURLToPath(import.meta.url) === process.argv[1];

if (isMainModule) {
  main().catch((error) => {
    console.error(error);
    process.exit(1);
  });
}

File: python/mcp_client.py
import sys
import asyncio
import json
from pathlib import Path
from pydantic import AnyUrl, FileUrl
from typing import Optional, Any
from contextlib import AsyncExitStack
from mcp import ClientSession, StdioServerParameters, types
from mcp.client.session import LoggingFnT, SamplingFnT
from mcp.client.stdio import stdio_client
# --- STREAMABLE HTTP transport ---
# Uncomment to switch (also comment out the "from mcp.client.stdio import stdio_client" line above).
# from mcp.client.streamable_http import streamablehttp_client
from mcp.shared.context import RequestContext
from mcp.shared.session import ProgressFnT
from mcp.types import ErrorData, ListRootsResult, Root


class MCPClient:
    def __init__(
        self,
        command: str,
        args: list[str],
        env: Optional[dict] = None,
        sampling_callback: Optional[SamplingFnT] = None,
        logging_callback: Optional[LoggingFnT] = None,
        roots: Optional[list[str]] = None,
    ):
        self._command = command
        self._args = args
        self._env = env
        self._sampling_callback = sampling_callback
        self._logging_callback = logging_callback
        self._roots = self._create_roots(roots) if roots else []
        self._session: Optional[ClientSession] = None
        self._exit_stack: AsyncExitStack = AsyncExitStack()

    def _create_roots(self, root_paths: list[str]) -> list[Root]:
        """Convert path strings to Root objects."""
        roots = []
        for path in root_paths:
            p = Path(path).resolve()
            file_url = FileUrl(f"file://{p}")
            roots.append(Root(uri=file_url, name=p.name or "Root"))
        return roots

    async def _handle_list_roots(
        self, context: RequestContext["ClientSession", Any]
    ) -> ListRootsResult | ErrorData:
        """Callback for when the server requests the client's roots."""
        return ListRootsResult(roots=self._roots)

    async def connect(self):
        # --- STDIO transport (default) ---
        server_params = StdioServerParameters(
            command=self._command,
            args=self._args,
            env=self._env,
        )
        stdio_transport = await self._exit_stack.enter_async_context(
            stdio_client(server_params)
        )
        _read, _write = stdio_transport

        # --- STREAMABLE HTTP transport ---
        # Comment out the stdio block above and uncomment the block below to switch.
        # Requires the server to be running with `mcp.run(transport="streamable-http", ...)`
        # (see mcp_server.py) and reachable at this URL.
        # http_transport = await self._exit_stack.enter_async_context(
        #     streamablehttp_client("http://127.0.0.1:8000/mcp")
        # )
        # _read, _write, _get_session_id = http_transport

        self._session = await self._exit_stack.enter_async_context(
            ClientSession(
                _read,
                _write,
                sampling_callback=self._sampling_callback,
                logging_callback=self._logging_callback,
                list_roots_callback=self._handle_list_roots
                if self._roots
                else None,
            )
        )
        await self._session.initialize()

    def session(self) -> ClientSession:
        if self._session is None:
            raise ConnectionError(
                "Client session not initialized or cache not populated. Call connect_to_server first."
            )
        return self._session

    async def list_tools(self) -> list[types.Tool]:
        result = await self.session().list_tools()
        return result.tools

    async def call_tool(
        self,
        tool_name: str,
        tool_input: dict,
        progress_callback: Optional[ProgressFnT] = None,
    ) -> types.CallToolResult | None:
        return await self.session().call_tool(
            tool_name, tool_input, progress_callback=progress_callback
        )

    async def list_prompts(self) -> list[types.Prompt]:
       result = await self.session().list_prompts()
       return result.prompts

    async def get_prompt(self, prompt_name, args: dict[str, str]):
        result = await self.session().get_prompt(prompt_name, args)
        return result.messages

    async def read_resource(self, uri: str) -> Any:
       result = await self.session().read_resource(AnyUrl(uri))
       resource = result.contents[0]

       if isinstance(resource, types.TextResourceContents):
        if resource.mimeType == "application/json":
            return json.loads(resource.text)

            return resource.text

    async def cleanup(self):
        await self._exit_stack.aclose()
        self._session = None

    async def __aenter__(self):
        await self.connect()
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        await self.cleanup()


# For testing
async def main():
    async with MCPClient(
        # If using Python without UV, update command to 'python' and remove "run" from args.
        command="uv",
        args=["run", "mcp_server.py"],
    ) as _client:
        result = await _client.list_tools()
        print()


if __name__ == "__main__":
    if sys.platform == "win32":
        asyncio.set_event_loop_policy(asyncio.WindowsProactorEventLoopPolicy())
    asyncio.run(main())
